buildSrc/src/main/groovy/net/minecraftforge/forge/tasks/checks/CheckATs.groovy

```diff
@@ -0,0 +1,304 @@
+
+            // Process Groups, this will remove any entries outside the group that is covered by the group
+            if (entry.group) {
+                final jcls = inheritance[entry.cls.replaceAll('\\.', '/')]
+                if (jcls === null) {
+                    itr.remove()
+                    reporter.report("Invalid group: $key")
+                } else if ('*' == entry['desc']) {
+                    if (!jcls.fields) {
+                        itr.remove()
+                        reporter.report("Invalid group, class has no fields: $key")
+                    } else {
+                        jcls.fields.each { field, value ->
+                            final fkey = entry.cls + ' ' + field
+                            if (accessLevel(value.access) < accessStr(entry.modifier)) {
+                                if (lines.containsKey(fkey)) {
+                                    toRemove.add(fkey)
+                                } else if (!entry.existing.contains(fkey)) {
+                                    println('Added: ' + fkey)
+                                }
+                                entry.children.add(fkey)
+                            } else if (lines.containsKey(fkey)) {
+                                toRemove.add(fkey)
+                                println('Removed: ' + fkey)
+                            }
+                        }
+                        entry.existing.findAll { it !in entry.children }.each { println('Removed: ' + it) }
+                    }
+                } else if ('*()' == entry.desc) {
+                    if (!jcls.methods) {
+                        itr.remove()
+                        reporter.report("Invalid group, class has no methods: $key")
+                    } else {
+                        jcls.methods.each { mtd, value ->
+                            if (mtd.startsWith('<clinit>') || mtd.startsWith('lambda$')) return
+                            key = entry.cls + ' ' + mtd.replace(' ', '')
+                            if (accessLevel(value.access) < accessStr(entry.modifier)) {
+                                if (lines.containsKey(key)) {
+                                    toRemove.add(key)
+                                } else if (!entry.existing.contains(key)) {
+                                    println('Added: ' + key)
+                                }
+                                entry.children.add(key)
+                            } else if (lines.containsKey(key)) {
+                                toRemove.add(key)
+                                println('Removed: ' + key)
+                            }
+                        }
+                        entry.existing.findAll { it !in entry.children }.each { println('Removed: ' + it) }
+                    }
+                } else if ('<init>' == entry.desc) { //Make all public non-abstract subclasses
+                    inheritance.each { tcls, value ->
```

Comment:
I wonder if this could be optimized by moving this outside of the AT entry checking loop, so it only needs to iterate over all the classes in the inheritance map once.